flang/lib/Lower/OpenMP.cpp

```diff
@@ -522,11 +549,96 @@ void DataSharingProcessor::defaultPrivatize() {
         !sym->GetUltimate().has<Fortran::semantics::NamelistDetails>() &&
         !symbolsInNestedRegions.contains(sym) &&
         !symbolsInParentRegions.contains(sym) &&
-        !privatizedSymbols.contains(sym)) {
+        !privatizedSymbols.contains(sym))
+      doPrivatize(sym);
+  }
+}
+
+void DataSharingProcessor::doPrivatize(const Fortran::semantics::Symbol *sym) {
+  if (!useDelayedPrivatization) {
+    cloneSymbol(sym);
+    copyFirstPrivateSymbol(sym);
+    return;
+  }
+
+  Fortran::lower::SymbolBox hsb = converter.lookupOneLevelUpSymbol(*sym);
+  assert(hsb && "Host symbol box not found");
+
+  mlir::Type symType = hsb.getAddr().getType();
+  mlir::Location symLoc = hsb.getAddr().getLoc();
+  std::string privatizerName = sym->name().ToString() + ".privatizer";
+  bool isFirstPrivate =
+      sym->test(Fortran::semantics::Symbol::Flag::OmpFirstPrivate);
+
+  mlir::omp::PrivateClauseOp privatizerOp = [&]() {
+    auto moduleOp = firOpBuilder.getModule();
+
+    auto uniquePrivatizerName = fir::getTypeAsString(
+        symType, converter.getKindMap(),
+        sym->name().ToString() +
```

Comment:
> Would it make sense to make the privatizer name only depend on the variable type and the data sharing type (private, firstprivate, ...)?

I did this initially as suggested by @clementval & @luporl and it works fine. However, in the `alloc` region of the privatizer, we have `binc_name` attributes that have the name of the first symbol with that type and data-sharing type. So if you `var1` and `var2` sharing the same type and data-sharing type, you get something like this for their shared privatizer:
```
fir.alloca i32 {bindc_name = "var1", pinned, uniq_name = "_QFdelayed_privatization_privateEvar1"}
```

So if this is fine, I can go back to not using the variable name. I am just not very familiar with the semantics of these name-related attributes so I went the safer way.

Reply:
I removed the symbol name part of the privatizer name since I got no objections/concerns about having `bindc_name` wrong.

Reply:
I see, I didn't realize about the `uniq_name` attribute here. I guess it's not great to have it incorrectly represent the original Fortran variable. In this case I think that probably the best solution is to make sure that the privatizer is unique to each variable, so rather than reusing privatizers that do the same thing we need to make sure to only reuse them when they refer to the same variable and not variables named the same with the same type and use defined in different functions.

So, my suggestion would be to make the privatizer name contain the information that is also used to produce the `uniq_name` attribute plus the data sharing and variable types, to make sure there won't be name collisions. At least adding the parent function name in addition to the variable name should be good for now, but maybe we'd have to throw in some reference to the Fortran module as well to avoid collisions there as well.

Reply:
I agree with you. I used `converter.mangleName(sym)` which returns the value used for `uniq_name`.